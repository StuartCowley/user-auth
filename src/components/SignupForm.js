import React, { useState } from 'react'

import '../styles/signup-form.css';

const SignupForm = ({setUser}) => {
  const initialFormValues = {
    firstName: "",
    lastName: "",
    email: "",
    password: ""
  };
  const [formValues, setFormValues] = useState(initialFormValues);

  const handleInputChange = (e) => {
    const updatedValues = { ...formValues };
    updatedValues[e.target.id] = e.target.value
    
    setFormValues(updatedValues);
  };

  const handleSubmit = () => {
    setUser(formValues);
  }

  return (
    <div className='signup-form'>
      <h1>Signup here:</h1>
      <div className='signup-form__element'>
        <label htmlFor="firstName">First Name</label>
        <input
          className="signup-form__input"
          type="text"
          id="firstName"
          onChange={handleInputChange} 
          value={formValues.firstName} 
        />
      </div>
      <div className='signup-form__element'>
        <label htmlFor="lastName">Last Name</label>
        <input
          className="signup-form__input"
          type="text"
          id="lastName"
          onChange={handleInputChange} 
          value={formValues.lastName} 
        />
      </div>
      <div className='signup-form__element'>
        <label htmlFor="email">Email</label>
        <input
          className="signup-form__input"
          type="email"
          id="email"
          onChange={handleInputChange} 
          value={formValues.email} 
        />
      </div>
      <div className='signup-form__element'>
        <label htmlFor="password">Password</label>
        <input
          className="signup-form__input"
          type="password"
          id="password"
          onChange={handleInputChange} 
          value={formValues.password} 
        />
      </div>
      <button onClick={handleSubmit}>Submit</button>
    </div>
  )
}

export default SignupForm;