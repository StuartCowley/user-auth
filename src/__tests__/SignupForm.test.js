import { fireEvent, render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from "react-test-renderer";
import SignupForm from '../components/SignupForm';

describe("Signup form component", () => {
  describe("User is logged in", () => {
    const validProps = {
      allUsers: [
        {
          firstName: "testFirst",
          lastName: "testSecond",
          email: "test@email.com",
          password: "asdfgh"
        }
      ],
      setCurrentUser: jest.fn(),
      currentUser: {
        firstName: "testFirst",
        lastName: "testSecond",
        email: "test@email.com",
        password: "asdfgh"
      }
    };

    it("renders already logged in message", () => {
      render(
        <Router>
          <SignupForm
              allUsers={validProps.allUsers}
              setAllUsers={validProps.setAllUsers}
              setCurrentUser={validProps.setCurrentUser}
              currentUser={validProps.currentUser}
            />
        </Router>
      );

      const title = screen.getByText("You are already logged in, testFirst!");

      expect(title).toBeInTheDocument;
    })
  })

  describe("User is not logged in", () => {
    const validProps = {
      allUsers: [],
      setAllUsers: jest.fn(),
      setCurrentUser: jest.fn(),
      currentUser: null
    };
    it("matches snapshot", () => {
      const generated = renderer
        .create(
          <Router>
            <SignupForm
              allUsers={validProps.allUsers}
              setAllUsers={validProps.setAllUsers}
              setCurrentUser={validProps.setCurrentUser}
              currentUser={validProps.currentUser}
            />
          </Router>
        ).toJSON();

      expect(generated).toMatchSnapshot();
    })

    it("renders title", () => {
      render(
        <Router>
          <SignupForm
            allUsers={validProps.allUsers}
            setAllUsers={validProps.setAllUsers}
            setCurrentUser={validProps.setCurrentUser}
            currentUser={validProps.currentUser}
          />
        </Router>
      )
      const title = screen.getByText(/Signup here:/);

      expect(title).toBeInTheDocument;
    })

    it("renders all form elements", () => {
      render(
        <Router>
          <SignupForm
            allUsers={validProps.allUsers}
            setAllUsers={validProps.setAllUsers}
            setCurrentUser={validProps.setCurrentUser}
            currentUser={validProps.currentUser}
          />
        </Router>
      )
      const firstNameField = screen.getByText("First Name");
      const lastNameField = screen.getByText("Last Name");
      const emailField = screen.getByText("Email");
      const passwordField = screen.getByText("Password");
      const signupButton = screen.getByText("Submit");

      expect(firstNameField).toBeInTheDocument;
      expect(lastNameField).toBeInTheDocument;
      expect(emailField).toBeInTheDocument;
      expect(passwordField).toBeInTheDocument;
      expect(signupButton).toBeInTheDocument;
    })

    it("shows signup failure message when registering existing user", async () => {
      const updatedProps = {
        allUsers: [
          {
            firstName: "testFirst",
            lastName: "testLast",
            email: "test@email.com",
            password: "qwerty"
          }
        ],
        setAllUsers: jest.fn(),
        setCurrentUser: jest.fn()
      };

      render(
        <Router>
          <SignupForm
            allUsers={updatedProps.allUsers}
            setAllUsers={updatedProps.setAllUsers}
            setCurrentUser={updatedProps.setCurrentUser}
            currentUser={validProps.currentUser}
          />
        </Router>
      )
      const firstNameField = screen.getByLabelText("First Name");
      const lastNameField = screen.getByLabelText("Last Name");
      const emailField = screen.getByLabelText("Email");
      const passwordField = screen.getByLabelText("Password");
      const signupButton = screen.getByText("Submit");

      fireEvent.change(firstNameField, { target: { value: updatedProps.allUsers[0].firstName } });
      fireEvent.change(lastNameField, { target: { value: updatedProps.allUsers[0].lastName } });
      fireEvent.change(emailField, { target: { value: updatedProps.allUsers[0].email } });
      fireEvent.change(passwordField, { target: { value: updatedProps.allUsers[0].password } });
      fireEvent.click(signupButton);

      const errorMessage = await screen.findByText("User already exists");

      expect(errorMessage).toBeInTheDocument;
    })

    it("shows login failure message when all fields not filled out on submit", async () => {
      render(
        <Router>
          <SignupForm
            allUsers={validProps.allUsers}
            setAllUsers={validProps.setAllUsers}
            setCurrentUser={validProps.setCurrentUser}
            currentUser={validProps.currentUser}
          />
        </Router>
      )
      const firstNameField = screen.getByLabelText("First Name");
      const lastNameField = screen.getByLabelText("Last Name");
      const passwordField = screen.getByLabelText("Password");
      const signupButton = screen.getByText("Submit");

      fireEvent.change(firstNameField, "testFirst");
      fireEvent.change(lastNameField, "testLast");
      fireEvent.change(passwordField, "testPassword");
      fireEvent.click(signupButton);

      const errorMessage = await screen.findByText("All fields not filled out");

      expect(errorMessage).toBeInTheDocument;
    })
  })
});