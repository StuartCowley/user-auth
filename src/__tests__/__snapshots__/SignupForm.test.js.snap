// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup form component User is not logged in matches snapshot 1`] = `
<div
  className="signup-form"
>
  <h1>
    Signup here:
  </h1>
  <div
    className="signup-form__element"
  >
    <label
      htmlFor="firstName"
    >
      First Name
    </label>
    <input
      className="signup-form__input"
      id="firstName"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="signup-form__element"
  >
    <label
      htmlFor="lastName"
    >
      Last Name
    </label>
    <input
      className="signup-form__input"
      id="lastName"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="signup-form__element"
  >
    <label
      htmlFor="email"
    >
      Email
    </label>
    <input
      className="signup-form__input"
      id="email"
      onChange={[Function]}
      type="email"
      value=""
    />
  </div>
  <div
    className="signup-form__element"
  >
    <label
      htmlFor="password"
    >
      Password
    </label>
    <input
      className="signup-form__input"
      id="password"
      onChange={[Function]}
      type="password"
      value=""
    />
  </div>
  <button
    onClick={[Function]}
  >
    Submit
  </button>
</div>
`;
